File: C-Sharp/01_Basics/10_file_handling.cs
// ─────────────────────────────────────────
//  FILE HANDLING
// ─────────────────────────────────────────

using System;
using System.IO;
using System.Text.Json;
using System.Collections.Generic;

class FileHandling
{
    static void Main()
    {
        // ── Write to file ─────────────────────────
        File.WriteAllText("example.txt", "Hello, World!\nLine 2\nLine 3");
        // Overwrites if exists, creates if not

        // Write multiple lines
        string[] lines = { "Line 1", "Line 2", "Line 3" };
        File.WriteAllLines("lines.txt", lines);

        // ── Read entire file ──────────────────────
        string content = File.ReadAllText("example.txt");
        Console.WriteLine(content);

        // ── Read all lines into array ─────────────
        string[] readLines = File.ReadAllLines("example.txt");
        foreach (var line in readLines)
            Console.WriteLine(line);

        // ── Append to file ────────────────────────
        File.AppendAllText("example.txt", "\nAppended line");

        // ── StreamWriter — more control ───────────
        using (var writer = new StreamWriter("stream.txt"))
        {
            writer.WriteLine("First line");
            writer.WriteLine("Second line");
            writer.Write("No newline at end");
        }   // file is closed when using block exits

        // ── StreamReader — line by line ───────────
        using (var reader = new StreamReader("stream.txt"))
        {
            string? line;
            while ((line = reader.ReadLine()) != null)
                Console.WriteLine(line);
        }

        // ── Check existence ───────────────────────
        Console.WriteLine(File.Exists("example.txt"));   // True
        Console.WriteLine(File.Exists("missing.txt"));   // False

        // ── File info ─────────────────────────────
        var info = new FileInfo("example.txt");
        Console.WriteLine(info.Name);             // example.txt
        Console.WriteLine(info.Length);           // size in bytes
        Console.WriteLine(info.Extension);        // .txt
        Console.WriteLine(info.DirectoryName);    // parent directory
        Console.WriteLine(info.CreationTime);
        Console.WriteLine(info.LastWriteTime);

        // ── Copy / Move / Delete ──────────────────
        File.Copy("example.txt", "copy.txt", overwrite: true);
        File.Move("copy.txt", "moved.txt");
        File.Delete("moved.txt");

        // ── Directory operations ──────────────────
        Directory.CreateDirectory("test_dir");
        Console.WriteLine(Directory.Exists("test_dir"));   // True

        // List files in a directory
        foreach (var file in Directory.GetFiles("."))
            Console.WriteLine(file);

        // List with pattern
        foreach (var file in Directory.GetFiles(".", "*.txt"))
            Console.WriteLine(file);

        // ── Path manipulation ─────────────────────
        string fullPath = Path.Combine("folder", "sub", "file.txt");
        Console.WriteLine(fullPath);                    // folder\sub\file.txt

        Console.WriteLine(Path.GetFileName(fullPath));  // file.txt
        Console.WriteLine(Path.GetFileNameWithoutExtension(fullPath)); // file
        Console.WriteLine(Path.GetExtension(fullPath)); // .txt
        Console.WriteLine(Path.GetDirectoryName(fullPath)); // folder\sub

        string temp = Path.GetTempFileName();
        Console.WriteLine(temp);   // temp file path

        Console.WriteLine(Directory.GetCurrentDirectory());  // working directory

        // ── JSON read/write ───────────────────────
        var person = new { Name = "Alice", Age = 25, City = "NYC" };
        string json = JsonSerializer.Serialize(person, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText("person.json", json);
        Console.WriteLine(json);

        // Deserialize
        string jsonBack = File.ReadAllText("person.json");
        using var doc = JsonDocument.Parse(jsonBack);
        Console.WriteLine(doc.RootElement.GetProperty("Name").GetString());

        // ── Cleanup ───────────────────────────────
        foreach (var f in new[] { "example.txt", "lines.txt", "stream.txt", "person.json" })
            if (File.Exists(f)) File.Delete(f);

        Directory.Delete("test_dir");
    }
}

File: Java/01_Basics/10_FileHandling.java
// ─────────────────────────────────────────
//  FILE HANDLING
// ─────────────────────────────────────────

import java.io.*;
import java.nio.file.*;
import java.util.*;

public class FileHandling {
    public static void main(String[] args) throws IOException {

        // ── Write to file (NIO — modern way) ──────
        Path path = Path.of("example.txt");
        Files.writeString(path, "Hello, World!\nLine 2\nLine 3\n");

        // Write list of lines
        Files.write(Path.of("lines.txt"),
            List.of("Line 1", "Line 2", "Line 3"));

        // Append
        Files.writeString(path, "Appended line\n", StandardOpenOption.APPEND);

        // ── Read entire file ──────────────────────
        String content = Files.readString(path);
        System.out.println(content);

        // Read all lines into list
        List<String> lines = Files.readAllLines(path);
        lines.forEach(System.out::println);

        // ── BufferedReader — line by line ─────────
        try (BufferedReader reader = Files.newBufferedReader(path)) {
            String line;
            while ((line = reader.readLine()) != null)
                System.out.println(line);
        }

        // ── BufferedWriter ────────────────────────
        try (BufferedWriter writer = Files.newBufferedWriter(Path.of("output.txt"))) {
            writer.write("First line");
            writer.newLine();
            writer.write("Second line");
        }

        // ── Check existence & file info ───────────
        System.out.println(Files.exists(path));           // true
        System.out.println(Files.exists(Path.of("x")));  // false
        System.out.println(Files.size(path));             // size in bytes
        System.out.println(Files.isDirectory(path));      // false

        // ── Copy / Move / Delete ──────────────────
        Files.copy(path, Path.of("copy.txt"), StandardCopyOption.REPLACE_EXISTING);
        Files.move(Path.of("copy.txt"), Path.of("moved.txt"), StandardCopyOption.REPLACE_EXISTING);
        Files.delete(Path.of("moved.txt"));
        Files.deleteIfExists(Path.of("nonexistent.txt")); // no exception if missing

        // ── Directory operations ──────────────────
        Path dir = Path.of("test_dir");
        Files.createDirectories(dir);                     // create including parents
        System.out.println(Files.isDirectory(dir));       // true

        // Create a file inside dir
        Files.writeString(dir.resolve("file1.txt"), "content");

        // List files in directory
        try (var stream = Files.list(dir)) {
            stream.forEach(System.out::println);
        }

        // List all .txt files recursively
        try (var stream = Files.walk(Path.of("."))) {
            stream.filter(p -> p.toString().endsWith(".txt"))
                  .forEach(System.out::println);
        }

        // ── Path manipulation ─────────────────────
        Path p = Path.of("folder", "sub", "file.txt");
        System.out.println(p);                        // folder\sub\file.txt
        System.out.println(p.getFileName());          // file.txt
        System.out.println(p.getParent());            // folder\sub
        System.out.println(p.getRoot());              // null (relative path)
        System.out.println(p.toAbsolutePath());       // full absolute path
        System.out.println(p.resolve("other.txt"));   // folder\sub\other.txt
        System.out.println(p.normalize());            // cleans up . and ..

        // ── File attributes ───────────────────────
        var attrs = Files.readAttributes(path, java.nio.file.attribute.BasicFileAttributes.class);
        System.out.println(attrs.creationTime());
        System.out.println(attrs.lastModifiedTime());
        System.out.println(attrs.size());

        // ── Scanner — read user input / file ──────
        try (Scanner scanner = new Scanner(path)) {
            while (scanner.hasNextLine())
                System.out.println(scanner.nextLine());
        }

        // ── PrintWriter — formatted output ────────
        try (PrintWriter pw = new PrintWriter("formatted.txt")) {
            pw.printf("Name: %s, Age: %d%n", "Alice", 25);
            pw.println("Second line");
        }

        // ── Cleanup ───────────────────────────────
        for (String f : new String[]{"example.txt", "lines.txt", "output.txt", "formatted.txt"})
            Files.deleteIfExists(Path.of(f));

        Files.deleteIfExists(dir.resolve("file1.txt"));
        Files.deleteIfExists(dir);
    }
}
